src/cider/nrepl.clj

```diff
@@ -1,53 +1,398 @@
-            [cider.nrepl.middleware.test]
-            [cider.nrepl.middleware.trace]
-            [cider.nrepl.middleware.track-state]
-            [cider.nrepl.middleware.undef]
-            [cider.nrepl.middleware.version]))
+  (:require [clojure.tools.nrepl.middleware :refer [set-descriptor!]]
+            [clojure.tools.nrepl.middleware.session :refer [session]]
+            [clojure.tools.nrepl.middleware.pr-values :refer [pr-values]]
+            [clojure.tools.nrepl.server :as nrepl-server]
+            [cider.nrepl.middleware.util.cljs :as cljs]
+            [cider.nrepl.middleware.pprint :as pprint]
+            [cider.nrepl.middleware.version]
+            [cider.nrepl.print-method]))
+
+(def DELAYS (atom nil))
+
+(defn- resolve-or-fail [sym]
+  (or (resolve sym)
+      (throw (IllegalArgumentException. (format "Cannot resolve %s" sym)))))
+
+(defmacro run-delayed-handler
+  "Make a delay of `fn-name` and place it in `DELAYS` atom at compile time.
+  Require and invoke the delay at run-time with arguments h and msg."
+  [fn-name h msg]
+  (let [ns  (symbol (namespace `~fn-name))
+        sym (symbol (name `~fn-name))]
+    (swap! DELAYS assoc sym
+           (delay
+             (require `~ns)
+             (resolve-or-fail `~fn-name)))
+    `(@(get @DELAYS '~sym) ~h ~msg)))
+
+(defmacro ^{:arglists '([name handler-fn descriptor]
+                        [name handler-fn handle descriptor])}
+  def-wrapper
+  "Define delayed middleware (e.g. wrap-xyz).
+  `handler-fn` is an unquoted name of a function that takes two arguments -
+  `handler` and `message`. It is called only when certain conditions are met as
+  expressed by the optional `handle` argument. `handle` can be either a function
+  or a set of ops (strings). When a function, it must take a `msg` and return
+  truthy value when `handler-fn` should run. When `handle` is missing,
+  `handle-fn` is called when :op of `msg` is in the set of keys of the :handles
+  slot of the `descriptor`. When `handle` is a set it should contain extra ops,
+  besides those in :handles, on which `handle-fn` is triggered. `descriptor` is
+  passed to nREPLs `set-descriptor!`."
+  [name handler-fn & [handle descriptor]]
+  (let [[descriptor handle] (if descriptor [descriptor handle] [handle descriptor])
+        handle (eval handle)
+        descriptor (eval descriptor)
+        cond (if (or (nil? handle) (set? handle))
+               (let [ops-set (into (-> descriptor :handles keys set) handle)]
+                 `(~ops-set (:op ~'msg)))
+               `(~handle ~'msg))
+        doc (or (:doc descriptor) "")]
+    (assert descriptor)
+    `(do
+       (defn ~name ~doc [~'h]
+         (fn [~'msg]
+           (if (and ~cond (not (:inhibit-cider-middleware ~'msg)))
+             (run-delayed-handler ~handler-fn ~'h ~'msg)
+             (~'h ~'msg))))
+       (set-descriptor! #'~name ~descriptor))))
+
+
+;;; Wrappers
+
+(def-wrapper wrap-debug cider.nrepl.middleware.debug/handle-debug
```

Comment:
I don't like that all the middleware definitions are dumped in a file used for utility functions.

Reply:
Not sure I understand. That is a declaration of the middleware files. We can rename it, but i think `nrepl.clj` is good enough. What utility functions? There are no other utilities there except `def-wrapper` which is used for defining wrappers.

Reply:
My bad. I just saw what the file originally just defined the default handler.